### PR TITLE
Add Gravity param to Base

### PR DESCRIPTION
**Pull request by patrick-tolosa**
Notes for the reviewer:

- I could not find a proper place to put AutoGravity, It has no definite place since it's a very unique Entity. Suggestions are welcomed.
- AutoGravity is not tree-shakable, you'll get all the static functions once you import it. It's very small but still worth noting. The alternative looked very confusing (in terms of imports)
- There are still some issues with the compilation tests, so not all of Gravity could be implemented, however the vast majority is included in this PR.

### CODE REVIEW

File: scripts/createEntrypoints.ts [hunk 1]
```diff
@@ -13,6 +13,9 @@ createEntryPointFromESMPath('actions', 'actions');
 // Create an Entry Point from everything under bundles/esm/constants -> dist/constants
 createEntryPointFromESMPath('constants', 'constants');
 
+// Create an Entry Point from everything under bundles/esm/params -> dist/params
```
Comment (maoznir) [inline scripts/createEntrypoints.ts:16]: Is this related to this pr?

File: src/constants/gravityObjects/GravityObjects.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  readonly gravityObject:GravityObject;
+  private weight:number;
+  constructor(gravityObject:GravityObject, weight?: number) {
+    this.weight = weight;
+    this.gravityObject = gravityObject;
+  }
+
+  toString():string {
+    if(this.weight) {
+      const avoidStr = this.weight <=-100 ? 'avoid' : this.weight;
```
Comment (maoznir) [inline src/constants/gravityObjects/GravityObjects.ts:22]: 1. what if weight===0?
2. I would put this logic in a separate function and return:
`${name}${getWeight()}

File: scripts/createEntrypoints.ts (hunk 1 above)
Comment (patrick-tolosa) [reply scripts/createEntrypoints.ts:16]: Yes it is, it ensures that the Gravity param is importable by the end user

`import Gravity from '@base/params/gravity'

File: src/constants/gravityObjects/GravityObjects.ts (hunk 2 above)
Comment (patrick-tolosa) [reply src/constants/gravityObjects/GravityObjects.ts:22]: 1. What about it really? do we want to skip it? Is cat_0 not a valid input?
2. You'd like to see a function called 'concatNameAndWEight'? 

We could move all this logic (entirely) into a "calculateName or something similar, but then toString() will just call this function., feels a bit off.

wdyt?

Comment (maoznir) [reply src/constants/gravityObjects/GravityObjects.ts:22]: 1. if(this.weight), will be false for weight===0, that's why i asked, what would you like to happen when weight=0?
2. I expect the void/anything else logic to get more complicated so i'd put the weight logic elsewhere, like calcWeight.

Comment (nitzanj) [reply src/constants/gravityObjects/GravityObjects.ts]: I'm not sure it's well-defined anywhere, but a weight is always a number, `getWeight()` will always return a number. And if the user typed a number, we should keep it as a number in the URL.

The `avoid` keyword is for the other way around - it's there to make the user's life easier. using `avoid` without the need to understand weights. And in that case, we keep the word `avoid` as is. This is a strictly _server-side alias_ that we're not suppose to mess with - we should not replace one for the other on the client-side.

In other words, whatever the user did (either called it with an int weight, or invoked the `avoid()` function), whatever appears in the URL is exactly that - either the weight or the word `avoid`. No need to process anything.
In terms of internal implementation that probably means that you should keep another attribute besides the weight integer, and when building the string look at it first, and if nothing is there use the weight (or nothing at all if weight is null too). 

TL;DR 
`expect(Gravity.auto(AutoGravity.object(GravityObjects.CAT).avoid()).toString()).toBe('g_auto:cat_avoid');`
`expect(Gravity.auto(AutoGravity.object(GravityObjects.CAT, 50)).toString()).toBe('g_auto:cat_50');`
`expect(Gravity.auto(AutoGravity.object(GravityObjects.CAT, -100)).toString()).toBe('g_auto:cat_-100');`

Comment (patrick-tolosa) [reply src/constants/gravityObjects/GravityObjects.ts]: Fixed and added to the tests.

File: src/constants/gravityObjects/GravityObjects.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  constructor(name?:string) {
+    this.name = name;
+  }
+
+  toString(): string {
+    return this.name;
+  }
+}
+
+export class AutoGravity {
```
Comment (asisayag2) [inline src/constants/gravityObjects/GravityObjects.ts:12]: Why is that located under `constants/...`?

File: src/actions/resize/aspectRatio.ts [hunk 4]
```diff
@@ -1,4 +1,4 @@
-import {ResizeAction} from "./Resize";
+import ResizeAction from "./ResizeAction";
```
[the remark this answers is not in the document]
Comment (patrick-tolosa) [reply src/actions/resize/aspectRatio.ts:1]: Spotted it while working, it caused an infinite loop on build time, so I fixed it in this PR,
Can be removed if critical (technically it's a bug fix)

File: src/constants/gravityObjects/GravityObjects.ts [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+  private shouldAvoid:boolean;
+
+  constructor(gravityObject:GravityObject, weight?: number | string) {
+    this.weight = weight;
+    this.gravityObject = gravityObject;
+    this.shouldAvoid = false;
+  }
+
+  private shouldAddWeight() {
+    return typeof this.weight === 'number' || typeof this.weight === 'string' || this.shouldAvoid;
```
[the remark this answers is not in the document]
Comment (patrick-tolosa) [reply src/constants/gravityObjects/GravityObjects.ts:24]: There are several perspectives here.

From our perspective, weight is always a number.
From our customer's perspective, weight can be a number or a string (think of databases saving things as strings)

On top of that, a customer can call `.avoid()`, which automatically adds the word `avoid` as weight. `auto:cat_avoid`


Also, accepting a string is a type of future proof for `auto:cat_newOption`

File: src/constants/gravityObjects/GravityObjects.ts [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    }
+  }
+
+  toString():string {
+    // Future proofing, in case we'd like to support some custom string in the future, or if data is coming from a DB
+    if(this.shouldAddWeight()) {
+      return `${this.getName()}_${this.getWeight()}`;
+    } else {
+      return this.getName();
+    }
```
[the remark this answers is not in the document]
Comment (patrick-tolosa) [reply src/constants/gravityObjects/GravityObjects.ts:45]: I guess, I personally prefer this syntax for readability, but open for change.

File: src/constants/gravityObjects/GravityObjects.ts [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+export const BIRD = new GravityObject('bird');
+export const DOG = new GravityObject('dog');
+export const CAT = new GravityObject('cat');
+export const MICROWAVE = new GravityObject('microwave');
+export const REFRIGERATOR = new GravityObject('refrigerator');
+export const BOTTLE = new GravityObject('bottle');
+export const SINK = new GravityObject('sink');
+export const SKATEBOARD = new GravityObject('skateboard');
+export const PERSON = new GravityObject('person');
```
[the remark this answers is not in the document]
Comment (patrick-tolosa) [reply src/constants/gravityObjects/GravityObjects.ts:77]: The Enums for GravityObjects, but I'm sure that's not what you meant.

File: __TESTS__/unit/types/Gravity.test.ts [hunk 8]
```diff
@@ -0,0 +1,70 @@
+import Gravity from "../../../src/params/gravity/Gravity";
+import * as GravityESM from "../../../src/params/gravity/Gravity";
+
+import * as GravityObjects from "../../../src/constants/gravityObjects/GravityObjects";
+import {AutoGravity} from "../../../src/constants/gravityObjects/GravityObjects";
+
+
+
+describe('Gravity Param', () => {
+  it('Ensure ESM and Default export the same thing', () => {
```
[the remark this answers is not in the document]
Comment (patrick-tolosa) [reply __TESTS__/unit/types/Gravity.test.ts:10]: Sure,

GravityESM contains all the possible imports from Gravity in a single Object.
Our users will usually want to import only what they need 

`import {object} from 'base/params/gravity'`.

However, if you want to import everything inside `params/gravity` and not list every single function, you can import with a `*`

What we test here is that everything that's exported in the default way can also be imported as an individual module.


It's possible that `import Gravity from '...';  Gravity.object` to exist, but 
`import {object} from '...'` to fail. 

Default import does not have to contain the same objects as Named imports
This test ensures that we didn't forget anything (And won't forget anything in the future).


As a reference, we've had issues with this in Cloudinary-core, where a named export would exist, but a default export would not (or vice verca)

File: src/constants/gravityObjects/GravityObjects.ts (hunk 3 above)
Comment (patrick-tolosa) [reply src/constants/gravityObjects/GravityObjects.ts:12]: It's a very good question (and part of my notes to the reviewer).

Imports in JS are not like imports in Java or PHP, we have to explicitly import it from somewhere, we don't have namespaces.


There's simply no logical place to put AutoGravity, as there are no similar Entities.
For example, if this was part of Gravity, the code would look like this:

`Gravity.auto(Gravity.AutoGravity.object(GravityObjects.CAT))` - Not ideal.

Comment (maoznir) [reply src/constants/gravityObjects/GravityObjects.ts:12]: I suggest that things that are related to gravity will be in a Gravity folder, in folders like params, gravityObjects etc

Comment (patrick-tolosa) [reply src/constants/gravityObjects/GravityObjects.ts:12]: That would indeed be great, but how would a user import it? 
Unlike JAVA we can't just import the 'Gravity' namespace.

We also need to support granular imports, a user should only be able to pull only the constants he wants.
It's a lot of things that really limit how flexible we can be with the file structure.

How do you see the whole solution?

File: src/actions/resize/aspectRatio.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (asisayag2) [reply src/actions/resize/aspectRatio.ts:1]: I always prefer separating PRs

File: src/constants/gravityObjects/GravityObjects.ts (hunk 3 above)
Comment (asisayag2) [reply src/constants/gravityObjects/GravityObjects.ts:12]: Sure that there are not other places that such animals exists? Something that will justify creating a separated folder  structure for? 
I would never look for it under `constants`

File: src/constants/gravityObjects/GravityObjects.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (asisayag2) [reply src/constants/gravityObjects/GravityObjects.ts:24]: So... why not 
`return this.weight!==null || this.shouldAvoid`
?

Another option is delegating the entire weight handling to the `getWeight` method.
In the `toString` just return `${this.getName()}${this.getWeight()}`
and in the `getWeight` return `'_avoid' / '' / '_${this.weight}` according to what is sent

File: src/constants/gravityObjects/GravityObjects.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (asisayag2) [reply src/constants/gravityObjects/GravityObjects.ts:77]: It's supposed to be a huge enum. maybe have it in its own file?
